### PR TITLE
implemented load model feature

### PR DESCRIPTION
**Pull request by jrego43**
- Added feature and test for loading model parameters via a state_dict
- Includes storing model parameters since this shares code with #16
- Very much WIP, especially the `set_weights()` and `set_biases()` convolution layer method implementations

### CODE REVIEW

File: src/layers/model.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+use std::fs::File;
+
+pub trait Model {
+    fn save(&self, path: &str) {
+        let mut file = File::create(path).unwrap();
+
+        let state_dict = self.state_dict();
+
+        ser::to_writer(&mut file, &state_dict, true).unwrap();
+    }
```
Comment (LazaroHurtado) [inline src/layers/model.rs:12]: @Demba3 you commented on this, how about naming it `save_state_dict`? Although it is different from Pytorch the method name explicitly denotes the functionality

File: examples/feedforward_nn.rs [hunk 2]
```diff
@@ -7,11 +7,11 @@ fn main() {
     let model = sequential!(
         Ix1,
         [
-            Linear::new(3, 4),
+            Linear::new(String::from("fc1"), 3, 4),
```
Comment (LazaroHurtado) [inline examples/feedforward_nn.rs:10]: I already mentioned this in another comment, but this can be changed to `Linear::new("fc1", 3, 4)` after implementing the changes I suggested

File: src/layers/linear.rs [hunk 3]
```diff
@@ -26,9 +31,47 @@ impl Layer<Ix1, Ix1> for Linear {
...
+
+    fn set_weights(&mut self, new_weights: Vec<f64>) {
+        if self.weights.len() != new_weights.len() {
+            panic!("Loaded weight count does not match layer weight count.");
+        }
+
+        self.weights
+            .iter_mut()
+            .zip(new_weights.iter())
+            .for_each(|(value_num, num)| *value_num = val!(*num));
```
Comment (LazaroHurtado) [inline src/layers/linear.rs:56]: To not have to dereference `num` you can do `zip(new_weights.into_iter())`. Here is my suggestion
```suggestion
        let mut iter = new_weights.into_iter();
        self.weights
            .map_inplace(|v| *v = val!(iter.next().unwrap()));
```

File: src/layers/linear.rs [hunk 4]
```diff
@@ -26,9 +31,47 @@ impl Layer<Ix1, Ix1> for Linear {
...
+    }
+
+    fn set_biases(&mut self, new_biases: Vec<f64>) {
+        if self.biases.len() != new_biases.len() {
+            panic!("Loaded bias count does not match layer bias count.");
+        }
+
+        self.biases
```
Comment (LazaroHurtado) [inline src/layers/linear.rs:64]: Same as above can be done here

File: src/layers/sequential.rs [hunk 5]
```diff
@@ -60,3 +61,32 @@ where
...
+        let mut state_dict: IndexMap<String, Vec<f64>> = IndexMap::new();
+        let layers_iter = self.layers.iter();
+
+        for layer in layers_iter.filter(|x| x.is_trainable()) {
+            let weights_key = format!("{}{}", layer.name(), ".weight");
+            let biases_key = format!("{}{}", layer.name(), ".bias");
+
+            state_dict.insert(weights_key, layer.weights().clone());
+            state_dict.insert(biases_key, layer.biases().clone());
+        }
```
Comment (LazaroHurtado) [inline src/layers/sequential.rs:76]: Might as well do the filtering in the same line as you converted the vector into an iterator. Also you don't need to clone weights and biases before inserting.
```suggestion
        let trainable_layers = self.layers.iter().filter(|x| x.is_trainable());

        for layer in trainable_layers {
            let weights_key = format!("{}{}", layer.name(), ".weight");
            let biases_key = format!("{}{}", layer.name(), ".bias");

            state_dict.insert(weights_key, layer.weights());
            state_dict.insert(biases_key, layer.biases());
        }
```

File: src/layers/sequential.rs [hunk 6]
```diff
@@ -60,3 +61,32 @@ where
...
+
+        state_dict
+    }
+
+    fn load_state_dict(&mut self, state_dict: &IndexMap<String, Vec<f64>>) {
```
Comment (LazaroHurtado) [inline src/layers/sequential.rs:81]: same as above

File: src/utils/kernel.rs [hunk 7]
```diff
@@ -49,6 +49,22 @@ where
...
+
+    pub fn weights(&self) -> Tensor<E> {
+        self.weights.clone()
+    }
+
+    pub fn bias(&self) -> f64 {
+        self.bias.value()
+    }
+
+    pub fn weights_mut(&mut self) -> &mut Tensor<E> {
```
Comment (LazaroHurtado) [inline src/utils/kernel.rs:61]: I feel like `weights_mut` and `set_bias` both attempt to do the same thing but are implemented in different ways, choosing one approach might be cleaner

File: tests/model.rs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    state_dict.insert(String::from("fc1.bias"), vec![0.1, -0.2, 0.3]);
+    state_dict.insert(
+        String::from("fc2.weight"),
+        vec![-0.1, 0.2, -0.3, -0.4, 0.5, -0.6],
+    );
+    state_dict.insert(String::from("fc2.bias"), vec![-0.4, 0.5]);
+
+    model.load_state_dict(&state_dict);
+
+    assert_eq!(model.state_dict(), state_dict);
```
Comment (LazaroHurtado) [inline tests/model.rs:34]: Might be nice to include a `assert_ne!` before loading the state dict to show that it actually was different

**Review by LazaroHurtado — CHANGES_REQUESTED** [review]
Thanks for taking on such a big task, I left some suggestions!

File: src/utils/kernel.rs (hunk 7 above)
Comment (jrego43) [reply src/utils/kernel.rs:61]: Agreed, these methods were some of the ones I was least confident in. How would you suggest having a single approach, though, since kernel weights is a value tensor, whereas kernel bias is a single value?

Comment (LazaroHurtado) [reply src/utils/kernel.rs:61]: By one method i mean either `weights_mut` and `bias_mut` or `set_weights` and `set_bias`

File: src/layers/layer.rs [hunk 9]
```diff
@@ -10,4 +10,24 @@ where
...
+    fn set_weights(&mut self, _new_weights: Vec<f64>) {}
+
+    fn set_biases(&mut self, _new_biases: Vec<f64>) {}
+
+    fn is_trainable(&self) -> bool {
+        false
+    }
+
+    fn name(&self) -> String {
+        String::from("DEFAULT_NAME")
```
[the remark this answers is not in the document]
Comment (jrego43) [reply src/layers/layer.rs:31]: Noted. Although in PyTorch you don't have to specify names for activation functions. Should we deviate from it in this instance?

File: src/layers/convolution.rs [hunk 10]
```diff
@@ -84,4 +104,44 @@ where
...
+        }
+
+        let mut new_iter = new_weights.iter();
+
+        self.kernels.iter_mut().for_each(|kernel| {
+            kernel
+                .weights_mut()
+                .iter_mut()
+                .for_each(|weight| *weight = val!(*new_iter.next().unwrap()))
+        });
```
[the remark this answers is not in the document]
Comment (jrego43) [reply src/layers/convolution.rs:126]: > Also changing the code I just suggested to (weight + 1.).into() will still pass all tests, is this expected?

Not sure what you mean here as I see no reason why it wouldn't pass all tests. Are you not just adding 1 to every weight?

File: src/utils/kernel.rs (hunk 7 above)
Comment (jrego43) [reply src/utils/kernel.rs:61]: Ok I see what you mean. Would it be fine to have something like the following in convolution.rs:

```
set_weights() {
  for each kernel in kernels {
    kernel.set_weights()
  }
}
```

I avoided this because of the repeated name `set_weights`, so I'm leaning towards `weights_mut` and `bias_mut` as you suggested.

File: src/layers/convolution.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (LazaroHurtado) [reply src/layers/convolution.rs:126]: It passed all tests because there is no test for checking if state dict for a Convolution layer was completed correctly.
Either way, what I was trying to get at was that if I added 1 to the weights in this method then there would be a disparity between the layers' weights and the state dict that was passed. But again, there was no test coverage for this so it makes sense why it passed

[the remark this answers is not in the document]
Comment (jrego43) [reply src/layers/convolution.rs:126]: Gotcha. Will add a test for convolutional layers.

**Comment by jrego43** [comment]
It's possible I missed a thing here or there, let me know!

File: src/activation/softmax.rs [hunk 11]
```diff
@@ -77,4 +77,8 @@ where
     fn forward(&self, input: &Tensor<D>) -> Tensor<D> {
         self.activate(input)
     }
+
+    fn name(&self) -> String {
+        String::from("softmax")
```
Comment (LazaroHurtado) [inline src/activation/softmax.rs:82]: Just to be consistent, should this be "Softmax" instead since other activation functions like ReLU are capitalized?

File: src/layers/convolution.rs [hunk 12]
```diff
@@ -73,15 +77,50 @@ where
         let mut single_channel_dim = <D>::zeros(D::NDIM.unwrap());
         let mut output_channels = vec![];
 
-        self.kernels.iter().for_each(|kernel| {
+        for kernel in self.kernels.iter() {
```
Comment (LazaroHurtado) [inline src/layers/convolution.rs:80]: Thank you!

File: src/activation/softmax.rs (hunk 11 above)
Comment (jrego43) [reply src/activation/softmax.rs:82]: Yep, missed that.

Comment (jrego43) [reply src/activation/softmax.rs:82]: It's a little strange because my local repo does have "Softmax" and I did add all changed files for the commit.

File: src/layers/sequential.rs [hunk 13]
```diff
@@ -33,10 +36,6 @@ where
             .collect()
     }
 
-    pub fn zero_grad(&self) {
-        self.parameters().iter().for_each(|value| value.zero_grad());
-    }
-
```
Comment (jrego43) [inline src/layers/sequential.rs:39]: I forgot I removed this when trying to make parameters() return a Vec<f64>, it's unused, but let me know if I should bring it back.

File: src/layers/linear.rs [hunk 14]
```diff
@@ -29,4 +35,32 @@ impl Layer<Ix1, Ix1> for Linear {
...
+        }
+    }
+
+    fn is_trainable(&self) -> bool {
+        true
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
```
Comment (LazaroHurtado) [inline src/layers/linear.rs:65]: I really wish these were able to be associated consts but doing so won't allow this trait to become a trait object :(

File: Cargo.toml [hunk 15]
```diff
@@ -12,6 +12,9 @@ rand_distr = "0.4"
 num-traits = "0.2"
 ndarray = { git = "https://github.com/LazaroHurtado/ndarray", branch = "windows_stride_feature" }
 ordered-float = "1.0.1"
+serde = "1.0"
+serde-pickle = "0.5.0"
```
Comment (LazaroHurtado) [inline Cargo.toml:16]: How come we are using `serde-pickle` version 0.5 over [1.1](https://docs.rs/serde-pickle/latest/serde_pickle/index.html)?

File: src/layers/sequential.rs (hunk 13 above)
Comment (LazaroHurtado) [reply src/layers/sequential.rs:39]: This is perfect, optimizers handle this logic now so it should have been removed a while ago, thank you!

File: src/layers/sequential.rs [hunk 16]
```diff
@@ -60,3 +59,46 @@ where
...
+            }
+        }
+
+        let trainable_layers = self.layers.iter_mut().filter(|x| x.is_trainable());
+
+        for layer in trainable_layers {
+            let (weight_key, bias_key) = (layer.name() + ".weight", layer.name() + ".bias");
+
+            layer.set_weights(state_dict.get(&weight_key).unwrap().clone());
+            layer.set_biases(state_dict.get(&bias_key).unwrap().clone());
```
Comment (LazaroHurtado) [inline src/layers/sequential.rs:101]: Can this logic be combined into one for-loop? Also is cloning the value of either the weight key or bias key needed?

```suggestion
        let trainable_layers = self.layers.iter_mut().filter(|x| x.is_trainable());

        for layer in trainable_layers {
            let (weight_key, bias_key) = (layer.name() + ".weight", layer.name() + ".bias");

            let saved_weights = state_dict.get(&weight_key).unwrap();
            if saved_weights.len() != layer.weights().len() {
                panic!("Wrong loaded weight count for layer \"{}\".", layer.name());
            }
            layer.set_weights(saved_weights.clone());
            
            let saved_bias = state_dict.get(&bias_key).unwrap();
            if saved_bias.len() != layer.biases().len() {
                panic!("Wrong loaded bias count for layer \"{}\".", layer.name());
            }
            layer.set_biases(saved_bias.clone());
```

File: src/utils/kernel.rs [hunk 17]
```diff
@@ -49,6 +49,22 @@ where
...
+        &mut self.weights
+    }
+
+    pub fn bias_mut(&mut self) -> &mut Value {
+        &mut self.bias
+    }
```
Comment (LazaroHurtado) [inline src/utils/kernel.rs:67]: Since I suggested using `.value_mut()` in a previous comment can we make do without these methods? Getting just the weights and bias and iterating over each value, although not mutable, should still work since `.value_mut()` does not require `&mut self`. Please lmk if this is possible

File: tests/model.rs [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    let path = "linear.pickle";
+    model1.save_state_dict(path);
+    model2.load_state_dict(path);
+
+    assert_eq!(model1.state_dict(), model2.state_dict());
+
+    assert!(
+        fs::remove_file(path).is_ok(),
+        "File \"linear.pickle\" could not be removed."
+    );
```
Comment (LazaroHurtado) [inline tests/model.rs:46]: Thank you for including this!

**Review by LazaroHurtado — COMMENTED** [review]
Awesome work!! left some small suggestions and also some questions I have :)

File: Cargo.toml (hunk 15 above)
Comment (jrego43) [reply Cargo.toml:16]: Leftover from Demba's PR, will change.

File: src/layers/convolution.rs [hunk 19]
```diff
@@ -73,15 +77,50 @@ where
...
+    fn biases(&self) -> Vec<f64> {
+        self.kernels.iter().map(|kernel| kernel.bias()).collect()
+    }
+
+    fn set_weights(&mut self, new_weights: Vec<f64>) {
+        let kernel_weights = new_weights.chunks_exact(self.weights().len() / self.kernels.len());
+
+        for (kernel, kernel_weight) in self.kernels.iter_mut().zip(kernel_weights) {
+            for (v, &weight) in kernel.weights_mut().iter_mut().zip(kernel_weight) {
+                *v = weight.into();
```
[the remark this answers is not in the document]
Comment (jrego43) [reply src/layers/convolution.rs:108]: Not all that familiar with `Rc` so I will take your word for it and do that.

File: src/layers/sequential.rs (hunk 16 above)
Comment (jrego43) [reply src/layers/sequential.rs:101]: You know what, you're right. I was thinking that counting must be done through a separate iterator first, because you could be loading a model halfway and then the counts mismatch, resulting in a corrupted model. However this is not at all an issue when considering that `panic!` will exit the program and drop all the allocated program memory anyway.

Comment (jrego43) [reply src/layers/sequential.rs:101]: Also, the clone is necessary because `set_weights` takes ownership of the passed `Vec<f64>`. But changing the signature so that it borrows the `Vec<f64>` instead does eliminate the need for clone, so I'll do that.

File: src/utils/kernel.rs (hunk 17 above)
Comment (jrego43) [reply src/utils/kernel.rs:67]: If I removed these methods, we still need replacements right? Because I can't access a kernel's `weights` or `bias` outside of the struct implementation. Calling `kernel.weights()` and `kernel.biases()` doesn't work because those return `Vec<f64>`.

Comment (LazaroHurtado) [reply src/utils/kernel.rs:67]: Good insight, might have to make those return tensors then!

File: src/layers/layer.rs [hunk 20]
```diff
@@ -5,9 +5,30 @@ where
...
     Out: Dimension,
 {
+    fn forward(&self, input: &Tensor<In>) -> Tensor<Out>;
+
     fn parameters(&self) -> Tensor<Ix1> {
+        let mut params = self.weights().into_raw_vec();
+        params.append(&mut self.biases().into_raw_vec());
+
+        Tensor::from_vec(params)
+    }
```
Comment (LazaroHurtado) [inline src/layers/layer.rs:15]: This is so nice to see, good job!

File: src/layers/linear.rs [hunk 21]
```diff
@@ -3,30 +3,57 @@ use crate::prelude::*;
...
-    fn parameters(&self) -> Tensor<Ix1> {
-        let mut params = self.weights.clone().into_raw_vec();
-        params.append(&mut self.biases.clone().into_raw_vec());
+    fn forward(&self, input: &Tensor<Ix1>) -> Tensor<Ix1> {
+        input.dot(&self.weights) + &self.biases
+    }
 
-        Tensor::from_vec(params)
+    fn weights(&self) -> Tensor<Ix1> {
+        Tensor::from_vec(self.weights.clone().into_raw_vec())
```
Comment (LazaroHurtado) [inline src/layers/linear.rs:33]: I think this operation is faster and safer

```suggestion
        self.weights.clone().into_shape(self.weights.len()).unwrap()
```

File: src/layers/linear.rs [hunk 22]
```diff
@@ -3,30 +3,57 @@ use crate::prelude::*;
...
     }
 
-    fn forward(&self, input: &Tensor<Ix1>) -> Tensor<Ix1> {
-        input.dot(&self.weights) + &self.biases
+    fn biases(&self) -> Tensor<Ix1> {
+        Tensor::from_vec(self.biases.clone().into_raw_vec())
```
Comment (LazaroHurtado) [inline src/layers/linear.rs:37]: bias is already a `Tensor<Ix1>`
```suggestion
        self.biases.clone()
```

File: src/utils/kernel.rs [hunk 23]
```diff
@@ -50,6 +50,14 @@ where
 
         Tensor::from_shape_vec(channel_shape, output).unwrap()
     }
+
+    pub fn weights(&self) -> Tensor<Ix1> {
+        Tensor::from_vec(self.weights.clone().into_raw_vec())
```
Comment (LazaroHurtado) [inline src/utils/kernel.rs:55]: Please reshape here as well

**Review by LazaroHurtado — CHANGES_REQUESTED** [review]
Great progress and thanks for cleaning up some out-of-scope files! Just some few tweaks left :D

File: src/layers/linear.rs (hunk 21 above)
Comment (LazaroHurtado) [reply src/layers/linear.rs:33]: My general thought process for these type of operations is, if the crate has an operation that does the same logic, use it since the developers probably optimized it

File: src/utils/kernel.rs (hunk 23 above)
Comment (LazaroHurtado) [reply src/utils/kernel.rs:55]:
```suggestion
        self.weights.clone().into_shape(self.weights.len()).unwrap()
```

**Review by LazaroHurtado — APPROVED** [review]
Amazing work, thank you for taking on this task and being so quick to reacting to reviews. Would also like to mention @Demba3 since this PR incorporates logic found in #16. I rebased to have only a single commit. This PR seems good to me, just need one more reviewer

**Review by Demba3 — APPROVED** [review]
LGTM 🚢🚢🚢
